presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java

```diff
@@ -776,22 +788,80 @@ public Page getNextPage()
                 if (qualifyingSet == null) {
                     qualifyingSet = new QualifyingSet();
                 }
-                qualifyingSet.setRange((int) currentGroupRowCount);
+                qualifyingSet.setRange(0, Math.min(ariaBatchRows, (int) currentGroupRowCount));
                 reader.setQualifyingSets(qualifyingSet, null);
+                if ((currentRowGroup & 0x3) == 0) {
+                    // Decay row size stats every 4 groups.
+                    numAriaRows /= 2;
+                    numAriaBytes /= 2;
+                }
                 if (reorderFilters && (currentRowGroup & 0x3) != 0 && currentRowGroup != 0) {
                     // Reconsider filter order every 4 row groups.
                     reader.maybeReorderFilters();
                 }
             }
-            qualifyingSet.clearTruncationPosition();
-            reader.advance();
+            if (qualifyingSet.isEmpty()) {
+                qualifyingSet.setRange(qualifyingSet.getEnd(), Math.min(qualifyingSet.getEnd() + ariaBatchRows, (int) currentGroupRowCount));
+            }
+            int bytesBeforeAdvance = reader.getResultSizeInBytes();
+            int numResultsBeforeAdvance = reader.getNumResults();
+            try {
+                reader.advance();
+            }
+            catch (BatchTooLargeException e) {
+                // The reader ran out of budget. Retry with a smaller
+                // input qualifying set. First remove any values added
+                // by the last advance.  The input qualifying set is
+                // unaltered when the reader throws this exception.
+                reader.compactValues(new int[0], numResultsBeforeAdvance, 0);
+                verify(ariaBatchRows > MIN_BATCH_ROWS, "Running out of budget with minimum batch size");
```

Comment:
This could happen if a particular row is very large. In this case, it should be a user error and not a system error. For user to be able to do something, the error message needs to include details such as batch size and budget. Furthermore, this check needs to be the very first statement, before reader.compactValues. Finally, I think the check should be using qualifyingSet.getEnd() - qualifyingSet.getBegin() and not ariaBatchRows.